src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/BusinessEvent/DispatchProductModelCreatedAndUpdatedEventSubscriber.php

```
@@ -0,0 +1,102 @@
+
+use Akeneo\Pim\Enrichment\Component\Product\Message\ProductModelCreated;
+use Akeneo\Pim\Enrichment\Component\Product\Message\ProductModelUpdated;
+use Akeneo\Pim\Enrichment\Component\Product\Model\ProductModelInterface;
+use Akeneo\Platform\Component\EventQueue\Author;
+use Akeneo\Platform\Component\EventQueue\BulkEvent;
+use Akeneo\Tool\Component\StorageUtils\StorageEvents;
+use Akeneo\UserManagement\Component\Model\UserInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\EventDispatcher\GenericEvent;
+use Symfony\Component\Messenger\MessageBusInterface;
+use Symfony\Component\Security\Core\Security;
+
+/**
+ * @copyright 202O Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class DispatchProductModelCreatedAndUpdatedEventSubscriber implements EventSubscriberInterface
+{
+    private Security $security;
+    private MessageBusInterface $messageBus;
+    private int $maxBulkSize;
+
+    /** @var array<ProductModelCreated|ProductModelUpdated> */
+    private array $events = [];
+
+    public function __construct(Security $security, MessageBusInterface $messageBus, int $maxBulkSize)
+    {
+        $this->security = $security;
+        $this->messageBus = $messageBus;
+        $this->maxBulkSize = $maxBulkSize;
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            StorageEvents::POST_SAVE => 'createAndDispatchProductModelEvents',
+            StorageEvents::POST_SAVE_ALL => 'dispatchBufferedProductModelEvents',
+        ];
+    }
+
+    public function createAndDispatchProductModelEvents(GenericEvent $postSaveEvent): void
+    {
+        /** @var ProductModelInterface */
+        $productModel = $postSaveEvent->getSubject();
+        if (false === $productModel instanceof ProductModelInterface) {
+            return;
```

Do we need to throw an exception here in CXP-563 ? If yes, please add a "TODO" comment.

---

No we don't need to :) this is a standard behavior of any event subscribers on a generic event (and post save event subscriber).
We need to validate the type of the "subject" object to not handle Attributes, simple Product, etc...